### PR TITLE
[Model Runner V2] support apply penalty for spec decode

### PR DESCRIPTION
**Pull request by izhuhaoran**
## Purpose

As titled, this pr support penalty for spec decode in MRV2

### CODE REVIEW

File: vllm/v1/worker/gpu/sample/penalties.py [hunk 1]
```diff
@@ -184,6 +210,148 @@ def apply_penalties(
...
+            # Store back logits
+            tl.store(
+                logits_ptr + logit_idx * logits_stride + vocab_offset, logits, mask=mask
+            )
+
+            # Update cumulative draft_counts for next position
+            if pos_in_req < num_logits - 1:
+                draft_token = tl.load(token_ids_ptr + logit_idx)
+                token_match = vocab_offset == draft_token
+                draft_counts = draft_counts + token_match.to(tl.int32)
```
[the remark this answers is not in the document]
Comment (izhuhaoran) [reply vllm/v1/worker/gpu/sample/penalties.py]: nice catch and fixed this.

**Review by WoosukKwon — COMMENTED** [review]
Thanks for the PR. QQ: Why don't we have `num_logits` thread blocks apply to each logits vector in parallel (instead of using `num_reqs` thread blocks)? IIUC, that can simplify the code quite a bit.

**Review by WoosukKwon — APPROVED** [review]
LGTM. Thanks for the great work!
